architecture/replication-layer.md

```diff
@@ -0,0 +1,114 @@
+
+When failures happen, though, CockroachDB automatically realizes nodes have stopped responding and work to redistribute your data to continue maximizing survivability. This process also works the other way around: when new nodes join your cluster, data automatically rebalances onto it, ensuring your load is evenly distributed.
+
+### Interactions with Other Layers
+
+In relationship to other layers in CockroachDB, the Distribution Layer:
+
+- Receives requests from and sends responses to the Distribution Layer.
+- Writes accepted requests to the Storage Layer.
+
+## Components
+
+### Raft
+
+Raft is a consensus protocol––an algorithm which makes sure that your data is safely stored on multiple machines, and that those machines agree on the current state even if some of them are temporarily disconnected.
+
+Raft organizes all nodes that contain a replica of a range into a group––unsurprisingly called a Raft Group.
+
+Raft elects a relatively long-lived leader which must be involved to propose commands. It heartbeats followers periodically and keeps their logs replicated. In the absence of heartbeats, followers become candidates after randomized election timeouts and proceed to hold new leader elections. Cockroach weights random timeouts such that the replicas with shorter round trip times to peers are more likely to hold elections first (not implemented yet). 
+
+Once a node receives a `BatchRequest` for a range it contains, it converts those KV operations into Raft commands. Those commands are proposed to the Raft group leader––which is what makes it ideal for the [Lease Holder](#leases) and the Raft leader to be one in the same––and written to the Raft log.
+
+Proposed Raft commands are proposed to have them serialized through quorum voting into a coherent distributed log.
+
+For a great overview of Raft, we recommend [The Secret Lives of Data](http://thesecretlivesofdata.com/raft/).
+
+#### Raft Logs
+
+When write receive a quorum, and are committed by the Raft group leader, they're appended to the Raft log. This provides an ordered set of commands that the replicas agreed on and is essentially the source of truth for consistent replication.
+
+Because this log is treated as serializable, it can be replayed to bring a node from a past state to its current state. This log also lets nodes that temporarily went offline to be "caught up" to the current state without needing to receive a copy of the existing data in the form of a snapshot.
+
+### Snapshots
+
+Each replica can be "snapshotted", which copies all of its data as of a specific timestamp (which is available because of [MVCC](storage-layer.html#mvcc)). This snapshot can be sent to other nodes during a rebalance event to expedite replication.
```

Comment:
This helps me understand this concept and why it's helpful, but I'm not clear on when and why a replica is "snapshotted". We can leave this for a follow-up PR, though, if you don't already have this in you head.